bundle/edu.gemini.itc.shared/src/main/scala/edu/gemini/itc/shared/ConfigExtractor.scala

```diff
@@ -0,0 +1,214 @@
+    import Gsaoi._
+    for {
+      filter      <- extract[Filter]        (c, FilterKey)
+      readMode    <- extract[ReadMode]      (c, ReadModeKey)
+    } yield {
+      val gems = GemsParameters(0.3, "K") // TODO: gems
+      GsaoiParameters(filter, readMode, gems)
+    }
+  }
+
+  private def extractNiri(targetEnv: TargetEnvironment, probe: GuideProbe, c: Config): String \/ NiriParameters = {
+    import Niri._
+    for {
+      filter      <- extract[Filter]        (c, FilterKey)
+      grism       <- extract[Disperser]     (c, DisperserKey)
+      camera      <- extract[Camera]        (c, CameraKey)
+      readMode    <- extract[ReadMode]      (c, ReadModeKey)
+      wellDepth   <- extract[WellDepth]     (c, WellDepthKey)
+      mask        <- extract[Mask]          (c, MaskKey)
+      altair      <- extractAltair          (targetEnv, probe, c)
+    } yield NiriParameters(filter, grism, camera, readMode, wellDepth, mask, altair)
+  }
+
+  private def extractAltair(targetEnv: TargetEnvironment, probe: GuideProbe, c: Config): String \/ Option[AltairParameters] = {
+    import AltairParams._
+
+    def altairIsPresent =
+      c.containsItem(AoSystemKey) && extract[String](c, AoSystemKey).rightMap("Altair".equals).getOrElse(false)
+
+    def extractGroup =
+      targetEnv.getPrimaryGuideProbeTargets(probe).asScalaOpt.fold("No guide star selected".left[GuideProbeTargets])(_.right)
+
+    def extractGuideStar(targets: GuideProbeTargets) =
+      targets.getPrimary.asScalaOpt.fold("No guide star selected".left[SPTarget])(_.right)
+
+    def extractMagnitude(guideStar: ITarget) = {
+      val r  = guideStar.getMagnitude(Band.r)
+      val R  = guideStar.getMagnitude(Band.R)
+      val UC = guideStar.getMagnitude(Band.UC)
+      if      (r.isDefined)  r.getValue.getBrightness.right
+      else if (R.isDefined)  R.getValue.getBrightness.right
+      else if (UC.isDefined) UC.getValue.getBrightness.right
+      else "No r, R or UC magnitude defined for guide star".left
+    }
+
+    if (altairIsPresent) {
+      for {
+        group     <- extractGroup
+        guideStar <- extractGuideStar(group)
+        magnitude <- extractMagnitude(guideStar.getTarget)
+        fieldLens <- extract[FieldLens]     (c, AoFieldLensKey)
+        wfsMode   <- extract[GuideStarType] (c, AoGuideStarTypeKey)
+      } yield {
+        val separation = distance(targetEnv.getBase, guideStar)
+        Some(AltairParameters(separation, magnitude, fieldLens, wfsMode))
+      }
+    } else {
+      None.right
+    }
+  }
+
+  private def extractObservingWavelength(c: Config): String \/ Double =
+    // Note: observing wavelength will only be available if instrument is configured for spectroscopy
+    if (c.containsItem(ObsWavelengthKey)) extractDoubleFromString(c, ObsWavelengthKey) else 0.0.right
+
+  // Extract a value of the given type from the configuration
+  private def extract[A](c: Config, key: ItemKey)(implicit clazz: ClassTag[A]): String \/ A =
```

Comment:
Nice, @swalker2m Should we promote these methods to be usable from anywhere on the OCS?